substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -205,10 +219,100 @@ private void abortIfSealed() {
+        // If the serialization target is primitive, it needs to get boxed, because the target is
+        // always an Object.
+        if (clazz.isPrimitive()) {
+            Class<?> boxedType = null;
+            switch (targetClassName) {
+                case "int":
+                    boxedType = Integer.class;
+                    break;
+                case "float":
+                    boxedType = Float.class;
+                    break;
+                case "double":
+                    boxedType = Double.class;
+                    break;
+                case "long":
+                    boxedType = Long.class;
+                    break;
+                case "byte":
+                    boxedType = Byte.class;
+                    break;
+                case "char":
+                    boxedType = Character.class;
+                    break;
+                case "boolean":
+                    boxedType = Boolean.class;
+                    break;
+                case "short":
+                    boxedType = Short.class;
+                    break;
+                case "void": // no need to register void for serialization
+                    return;
+                default:
+                    VMError.shouldNotReachHere(targetClassName + " is not primitive type.");
+            }
+            registerAllAssociatedClasses(condition, boxedType);
+            return;
+        } else if (!Serializable.class.isAssignableFrom(clazz)) {
+            println("Warning: The given class " + targetClassName + " is not serializable, it wouldn't get registered.\n");
+            return;
+        } else if (access.findSubclasses(clazz).size() > 1) {
+            println("Warning: The given class " + targetClassName +
+                            " has subclasses. We can't figure out which one should be registered. Skip.\n");
+            return;
+        }
+        try {
+            clazz.getDeclaredMethod("writeObject", ObjectOutputStream.class);
+            println("Warning: The given class " + targetClassName +
+                            " has customized serialization method writeObject. " +
+                            "We can't figure out what should be registered. Skip.\n");
+            return;
+        } catch (NoSuchMethodException e) {
+            // Expected case. Do nothing
+        }
+        register(condition, clazz);
+
+        if (clazz.isArray()) {
+            registerAllAssociatedClasses(condition, clazz.getComponentType());
+            return;
+        }
+        ObjectStreamClass osc = ObjectStreamClass.lookup(clazz);
+        try {
+            Method getDataLayoutMethod = ReflectionUtil.lookupMethod(ObjectStreamClass.class, "getClassDataLayout");
+            Object[] dataLayout = (Object[]) getDataLayoutMethod.invoke(osc);
+            for (int i = 0; i < dataLayout.length; i++) {
+                Class<?> classDataSlotClazz = Class.forName("java.io.ObjectStreamClass$ClassDataSlot");
```

Comment:
It looks like dynamic class loading is not allowed here (not sure why since this is hosted code though):

```
/home/runner/work/graal/graal/substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java:289: do not use dynamic class loading
```
source: https://github.com/oracle/graal/pull/3760/checks#step:9:1769

Judging by https://github.com/oracle/graal/pull/3760/files#diff-e7a9f15b46007ba6515c0e82b0d80801cbfcd01951b2ff13161bd3d8481fa807R200 I think the right way to do this is:
```suggestion
                Class<?> classDataSlotClazz = access.findClassByName("java.io.ObjectStreamClass$ClassDataSlot");
```

Reply:
Fixed as suggested